### PR TITLE
feat: Add `terragrunt_validate_inputs` hook to check unused and undefined inputs

### LINKED ISSUE #675: Terragrunt validate-inputs hook?
### What problem are you facing?

It looks like Terragrunt supports the validate-inputs subcommand. This would be helpful to run as a pre-commit hook to catch potential misconfigurations before runtime. It can also catch unused inputs with an additional flag that can help clean up configs.

https://terragrunt.gruntwork.io/docs/reference/cli-options/#validate-inputs

### How could pre-commit-terraform help solve your problem?

It looks like some time and effort can be saved by catching these config issues early in the development process.

### PR DESCRIPTION
**Pull request by jmreicha**
Put an `x` into the box if that apply:

- [ ] This PR introduces breaking change.
- [ ] This PR fixes a bug.
- [x] This PR adds new functionality.
- [ ] This PR enhances existing functionality.

### Description of your changes

Adds the ability to define a new `terragrunt_validate_inputs` hook for checking for unused and undefined variables in terragrunt.hcl configuration files.

Note: This does seem to require auth. I haven't found a way to get the hook to work without first authenticating to the cloud provider I was testing with. Please let me know if there is a way around this.

Fixes #675

### How can we test changes

Add the new pre-commit config:

```yaml
- repo: https://github.com/jmreicha/pre-commit-terraform
    rev: tg-validate
    hooks:
      - id: terragrunt_validate_inputs
        args:
          - --args=--terragrunt-strict-validate
```

Intentionally introduce a misconfiguration in a terragrunt.hcl configuration.

```hcl
inputs = {
	...
	foo = "bar"
}
```

Run the pre-comit hook and note the failure.

```bash
Terragrunt validate inputs...............................................Failed
- hook id: terragrunt_validate_inputs
- exit code: 1

WARN[0030] The following inputs passed in by terragrunt are unused:  prefix=[/home/jmreicha/github.com/lytx/cloudops-terragrunt/dataplatform/stage/us-west-2/redshift/ore-dp-non-prod-redshift-cluster]
WARN[0030]      - foo                                        prefix=[/home/jmreicha/github.com/lytx/cloudops-terragrunt/dataplatform/stage/us-west-2/redshift/ore-dp-non-prod-redshift-cluster]
WARN[0030]                                               prefix=[/home/jmreicha/github.com/lytx/cloudops-terragrunt/dataplatform/stage/us-west-2/redshift/ore-dp-non-prod-redshift-cluster]
INFO[0030] All required inputs are passed in by terragrunt  prefix=[/home/jmreicha/github.com/lytx/cloudops-terragrunt/dataplatform/stage/us-west-2/redshift/ore-dp-non-prod-redshift-cluster]
ERRO[0030] Terragrunt configuration has misaligned inputs. Strict mode enabled: true.
ERRO[0030] Unable to determine underlying exit code, so Terragrunt will exit with error code 1
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -85,10 +85,9 @@ If you want to support the development of `pre-commit-terraform` and [many other
 * [`tfupdate`](https://github.com/minamijoyo/tfupdate) required for `tfupdate` hook
 * [`hcledit`](https://github.com/minamijoyo/hcledit) required for `terraform_wrapper_module_for_each` hook
 
-
 #### 1.1 Custom Terraform binaries and OpenTofu support
 
-It is possible to set custom path to `terraform` binary.  
+It is possible to set custom path to `terraform` binary.
```
Comment (yermulnik) [inline README.md:90]: Please put the two trailing dots back. They denote line feed in markdown.

File: README.md [hunk 2]
```diff
@@ -113,7 +112,7 @@ All available tags [here](https://github.com/antonbabenko/pre-commit-terraform/p
 **Build from scratch**:
 
 > [!IMPORTANT]
-> To build image you need to have [`docker buildx`](https://docs.docker.com/build/install-buildx/) enabled as default builder.  
+> To build image you need to have [`docker buildx`](https://docs.docker.com/build/install-buildx/) enabled as default builder.
```
Comment (yermulnik) [inline README.md:115]: Please put the two trailing dots back. They denote line feed in markdown.

File: README.md [hunk 3]
```diff
@@ -219,7 +216,7 @@ Otherwise, you can follow [this gist](https://gist.github.com/etiennejeanneaurev
 
 Ensure your PATH environment variable looks for `bash.exe` in `C:\Program Files\Git\bin` (the one present in `C:\Windows\System32\bash.exe` does not work with `pre-commit.exe`)
 
-For `checkov`, you may need to also set your `PYTHONPATH` environment variable with the path to your Python modules.  
+For `checkov`, you may need to also set your `PYTHONPATH` environment variable with the path to your Python modules.
```
Comment (yermulnik) [inline README.md:219]: Please put the two trailing dots back. They denote line feed in markdown.

File: README.md [hunk 4]
```diff
@@ -336,8 +334,9 @@ You can specify environment variables that will be passed to the hook at runtime
 
 > [!IMPORTANT]
 > Variable values are exported _verbatim_:
-> - No interpolation or expansion are applied
-> - The enclosing double quotes are removed if they are provided
+>
```
Comment (yermulnik) [inline README.md:337]: Any reason for the extra line?

File: README.md [hunk 5]
```diff
@@ -372,10 +371,10 @@ Less verbose log levels will be implemented in [#562](https://github.com/antonba
 
 ### Many hooks: Parallelism
 
-> All, except deprecated hooks: `checkov`, `terraform_docs_replace` and hooks which can't be paralleled this way: `infracost_breakdown`, `terraform_wrapper_module_for_each`.  
+> All, except deprecated hooks: `checkov`, `terraform_docs_replace` and hooks which can't be paralleled this way: `infracost_breakdown`, `terraform_wrapper_module_for_each`.
```
Comment (yermulnik) [inline README.md:374]: Please put the two trailing dots back. They denote line feed in markdown.

File: README.md [hunk 6]
```diff
@@ -372,10 +371,10 @@ Less verbose log levels will be implemented in [#562](https://github.com/antonba
...
 > Also, there's a chance that parallelism have no effect on `terragrunt_fmt` and `terragrunt_validate` hooks
 
-By default, parallelism is set to `number of logical CPUs - 1`.  
+By default, parallelism is set to `number of logical CPUs - 1`.
```
Comment (yermulnik) [inline README.md:377]: Please put the two trailing dots back. They denote line feed in markdown.

File: README.md [hunk 7]
```diff
@@ -424,14 +422,12 @@ If you'd like to set parallelism value relative to number of CPU logical cores -
...
 
-
-
 ```yaml
 args:
   - --hook-config=--parallelism-ci-cpu-cores=N
 ```
 
-If you don't see code above in your `pre-commit-config.yaml` or logs - you don't need it.  
+If you don't see code above in your `pre-commit-config.yaml` or logs - you don't need it.
```
Comment (yermulnik) [inline README.md:430]: Please put the two trailing dots back. They denote line feed in markdown.
Please do it for the rest of the changes in this file that are not related to your change directly.

**Review by yermulnik — COMMENTED** [review]
LGTM (apart from unrelated formatting changes to README — please revert those unrelated changes, that were presumably applied automatically by your IDE)

ps: you may try and add the below to the config file of your VSCode to stop it from removing trailing dots from markdown
```json
    "[markdown]": {
        "files.trimTrailingWhitespace": false
    }
```

**Comment by MaxymVlasov** [comment]
Or just install https://editorconfig.org/ plugin to your favorite editor. I'll add it to the contrib article

**Comment by yermulnik** [comment]
> Or just install [editorconfig.org](https://editorconfig.org/) plugin to your favorite editor. I'll add it to the contrib article

Ah, right! Thanks, Max. I always miss that we have `.editorconfig` in here =)

**Comment by MaxymVlasov** [comment]
Here we go: https://github.com/antonbabenko/pre-commit-terraform/blob/master/.github/CONTRIBUTING.md#required-tools-and-plugins-to-simplify-review-process

If you merge the last changes to your branch, you'll be able to use extensions settings too.

Sorry for the inconvenience, that should have been added a long time ago.

**Comment by jmreicha** [comment]
Thanks for comments, I’ll get the formatting fixed up.

**Comment by jmreicha** [comment]
I think I got reformatted properly 🤞

**Review by yermulnik — COMMENTED** [review]
LGTM
Leaving final review to @MaxymVlasov

File: README.md [hunk 8]
```diff
@@ -1121,6 +1123,26 @@ It invokes `terragrunt providers lock` under the hood and terragrunt [does its'
...
+```yaml
+- id: terragrunt_validate_inputs
+  name: Terragrunt validate inputs
+  args:
+    # Optionally check for unused inputs
+    - --args=--terragrunt-strict-validate
+```
+
+> [!NOTE]
+> This hooks requires authentication to a given account if defined by config in order to work properly.
```
Comment (MaxymVlasov) [inline README.md:1144]: Can we provide here link to docs about authentication or provide more details?

I not sure which authentification is mentioned here. valid creds in `~/.aws/` `~/.azure/` etc? Or if not Cloud account, which account then?

Are there any terragrunt docs or issues about that?

**Review by MaxymVlasov — APPROVED** [review]

**Review by yermulnik — APPROVED** [review]
LGTM

**Comment by antonbabenko** [comment]
This PR is included in version 1.92.0 :tada:
